src/nvim/screen.c

```diff
@@ -2769,84 +2793,102 @@ win_line (
           if (row == startrow
               + filler_lines
               ) {
-            long num;
-            char *fmt = "%*ld ";
+            if (text_sign != 0 && strcmp((char *)wp->w_p_scl, "number") == 0) {
+              // Draw the sign instead of the number.
+              int len = draw_sign(&p_extra_free, text_sign,
+                                  number_width(wp), 1);
+              if (len > 0) {
+                n_extra = len;
+                p_extra = p_extra_free;
+                c_extra = NUL;
+              }
+            } else {
+              long num;
+              char *fmt = "%*ld ";
 
-            if (wp->w_p_nu && !wp->w_p_rnu)
-              /* 'number' + 'norelativenumber' */
-              num = (long)lnum;
-            else {
-              /* 'relativenumber', don't use negative numbers */
-              num = labs((long)get_cursor_rel_lnum(wp, lnum));
-              if (num == 0 && wp->w_p_nu && wp->w_p_rnu) {
-                /* 'number' + 'relativenumber' */
-                num = lnum;
-                fmt = "%-*ld ";
+              if (wp->w_p_nu && !wp->w_p_rnu) {
+                // 'number' + 'norelativenumber'
+                num = (long)lnum;
+              } else {
+                // 'relativenumber', don't use negative numbers
+                num = labs((long)get_cursor_rel_lnum(wp, lnum));
+                if (num == 0 && wp->w_p_nu && wp->w_p_rnu) {
+                  // 'number' + 'relativenumber'
+                  num = lnum;
+                  fmt = "%-*ld ";
+                }
               }
+
+              snprintf((char *)extra, n_extra + 1, fmt, number_width(wp), num);
+              if (wp->w_skipcol > 0) {
+                for (p_extra = extra; *p_extra == ' '; p_extra++) {
+                  *p_extra = '-';
+                }
+              }
+              if (wp->w_p_rl) {
+                // reverse line numbers
+                rl_mirror(extra);
+              }
+              p_extra = extra;
+              c_extra = NUL;
             }
 
-            sprintf((char *)extra, fmt,
-                number_width(wp), num);
-            if (wp->w_skipcol > 0)
-              for (p_extra = extra; *p_extra == ' '; ++p_extra)
-                *p_extra = '-';
-            if (wp->w_p_rl)                         /* reverse line numbers */
-              rl_mirror(extra);
-            p_extra = extra;
-            c_extra = NUL;
-          } else
-            c_extra = ' ';
-          n_extra = number_width(wp) + 1;
+          }
+
           char_attr = win_hl_attr(wp, HLF_N);
-          // When 'cursorline' is set highlight the line number of
-          // the current line differently.
-          // TODO(vim): Can we use CursorLine instead of CursorLineNr
-          // when CursorLineNr isn't set?
-          if ((wp->w_p_cul || wp->w_p_rnu)
-              && lnum == wp->w_cursor.lnum) {
+
+          if (text_sign != 0 && strstr((char *)wp->w_p_scl, "number") != NULL) {
```

Comment:
General note: `win_line` is a hot path, so it needs to stay fast. It's probably negligible, or maybe compilers optimize this `strstr`, but the general practice in `win_line` is avoid the risk. 

Typical "Vim style" would be to check `(*wp->w_p_scl == 'n' && *wp->w_p_scl == 'u')`. But that's a bit much, rather store a boolean at the top of `win_line`.

For the same reason we might want to mark `draw_sign` as `inline` (it should also be `static`).

Reply:
If this is a concern, the value could be parsed into an enum by option.c at the time of setting, as is done for other "hot-path" string options.

Reply:
As far as I can tell, this code is run only once per call to `win_line` anyway, except in the case of wrapping lines. So I don't think putting it at the beginning of `win_line` would do much. For now, I did the char-wise check.

As for using an enum, I haven't found any such cases for other options. Do you have an example?